pystoned/CNLS.py

```diff
@@ -1,304 +1,224 @@
-
-                # cost model
-                if fun == "cost":
-
-                    def qconvex_rule(model, i, h):
-                        arow = x[i]
-                        if i == h:
-                            return Constraint.Skip
-                        return sum(model.b[i, j] * arow[j] for j in model.j) >= \
-                               sum(model.b[h, j] * arow[j] for j in model.j)
-
-                    model.qconvex = Constraint(model.i, model.h, rule=qconvex_rule, doc='convexity constraint')
-
-    return model
\ No newline at end of file
+# Import pyomp module
+from pyomo.environ import *
+from pyomo.core.expr.numvalue import NumericValue
+
+
+class CNLS:
+    def __init__(self, y, x, cet='addi', fun='prod', rts='vrs'):
+        # cet     = "addi" : Additive composite error term
+        #         = "mult" : Multiplicative composite error term
+        # fun     = "prod" : production frontier
+        #         = "cost" : cost frontier
+        # rts     = "vrs"  : variable returns to scale
+        #         = "crs"  : constant returns to scale
+
+        ## TODO(error/warning handling): Check the configuration of the model exist
+        self.x = x.tolist()
+        self.y = y.tolist()
+        self.cet = cet
+        self.fun = fun
+        self.rts = rts
+
+        if type(self.x[0]) != list:
+            self.x = [x.tolist()]
+
+        # Initialize the pyomo model
+        self.__model__ = ConcreteModel()
+        self.__model__.I = Set(initialize=range(len(self.y)))
+        self.__model__.J = Set(initialize=range(len(self.x[0])))
+
+        # Initialize the variables
+        self.__model__.alpha = Var(self.__model__.I, doc='alpha')
+        self.__model__.beta = Var(
+            self.__model__.I, self.__model__.J, bounds=(0.0, None), doc='beta')
+        self.__model__.epsilon = Var(self.__model__.I, doc='residuals')
+        self.__model__.frontier = Var(
+            self.__model__.I, bounds=(0.0, None), doc='estimated frontier')
+
+        # Setup the objective function and constraints for model
+        self.__model__.objective = Objective(
+            rule=self.__objective_rule(),
+            sense=minimize,
+            doc='objective function')
+        self.__model__.regression_rule = Constraint(
+            self.__model__.I,
+            rule=self.__regression_rule(),
+            doc='regression equation')
+        if self.cet == "mult":
+            self.__model__.log_rule = Constraint(
+                self.__model__.I,
+                rule=self.__log_rule(),
+                doc="regression function")
+        self.__model__.concav_convex_rule = Constraint(
+            self.__model__.I,
+            self.__model__.I,
+            rule=self.__concav_convex_rule(),
+            doc='concavity/convexity constraint')
+
+        # Optimize model
+        self.optimization_status = 0
+        self.problem_status = 0
+
+    def optimize(self, remote=False):
```

Comment:
Regarding the solvers in the StoNED model, we usually use two kinds of solvers (i.e., MOSEK/CPLEX for QP problem, MINOS/KNITRO for NLP problem) to estimate the model. So, the logical for solver selection should be:
if the user defines the model as a Multiplicative model (i.e., `cet=multi`), the model should be estimated by remote server NEOS with knitro; if the `cet=addi`, the model can be solved by local MOSEK.

Reply:
I see.
But, I was considered about letting user to choose solving remotely/locally.
Do you have recommended local solver for NLP/ remote solver for QP?
If no, maybe stick to your design with some information log to user.
At least, let them know the information about the solver.

Reply:
Sure, there are some free local NLP solvers such as IPOT. But I do not want to recommend it due to its unstable computing capacity. 

The best remote solver for QP is also MOSEK. So, we can simply replace 'knitro" with "MOSEK" provided by NEOS.
`        else:
            solver = SolverManagerFactory('neos')
            self.problem_status = solver.solve(
                self.__model__, tee=True, opt='mosek')
            self.optimization_status = 1`

Reply:
I thought IPOPT is not reliable too.
Since I don't have a good choice for local NLP solver, I removed local optimization.

All the optimization now take place remotely, using `neos` with `mosek` for `additive model`, with `knitro` for `multiplicative model`.
Maybe, we can use IPOPT as local solver for NLP with warning log for recommending remote optimization for users?

Reply:
Good idea. I would suggest we can provide a option for selecting the local solvers or remote. On the one hand, perhaps in the near future, the free local NLP solvers are available. On the other hand, compared with remote MOSEK, I would prefer to use the local MOSEK API based on the solver stability. After all, if the user can not access the internet or the NEOS server crashes in some time, the local one would be a better choice. 

SO,  we can use following the codes?

     def optimize(self, remote = False):
           """Optimize the function by requested method"""
           ## TODO(error/warning handling): Check problem status after optimization
           if remote == False:
               if self.cet == "addi":
                   solver = SolverFactory('mosek')
                   self.problem_status = solver.solve(self.__model__, tee=True)
                   self.optimization_status = 1

               eif self.cet == "mult":
                    print("Estimating the multiplicative model will be available in near future.") 
                    return False

            else:
               if self.cet == "addi":
                   opt = "mosek"

               eif self.cet == "mult":
                   opt = "knitro"
               
               solver = SolverManagerFactory('neos')
               self.problem_status = solver.solve( self.__model__, tee=True, opt=opt)
               self.optimization_status = 1`

Reply:
OK, it is already updated in optimize method now!
Thanks for your contribution.

Reply:
OK.